### PR TITLE
[PropertyInfo] Make ReflectionExtractor compatible with ReflectionType changes in PHP 7.1

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q | A |
| --- | --- |
| Branch? | 2.8 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | https://github.com/symfony/symfony/issues/19677#issuecomment-244671828 |
| License | MIT |
| Doc PR | N/A |

### CODE REVIEW

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php [hunk 1]
```diff
@@ -231,7 +231,7 @@ private function extractFromAccessor($class, $property)
      */
     private function extractFromReflectionType(\ReflectionType $reflectionType)
     {
-        $phpTypeOrClass = (string) $reflectionType;
+        $phpTypeOrClass = method_exists($reflectionType, 'getName') ? $reflectionType->getName() : (string) $reflectionType;
```
Comment (dunglas) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: Why this change?

Comment (iltar) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: `\ReflectionType` does't have a `getName()` method according to the docs: http://php.net/manual/en/class.reflectiontype.php. Will this be added in the future perhaps?

File: src/Symfony/Component/PropertyInfo/Tests/Fixtures/Php71Dummy.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\PropertyInfo\Tests\Fixtures;
+
+/**
+ * @author Kévin Dunglas <dunglas@gmail.com>
```
Comment (iltar) [inline src/Symfony/Component/PropertyInfo/Tests/Fixtures/Php71Dummy.php]: Wrong author

File: src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php (hunk 1 above)
Comment (teohhanhui) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: Because `__toString` has changed in PHP 7.1 and will change in the future (it is not properly part of the BC promise). As advised in the PHP 7.1 UPGRADING file, we need to use `ReflectionNamedType::getName` going forward.

Comment (teohhanhui) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: @iltar

> ReflectionType::__toString() will now return the type name with a leading
> "?" if it is nullable. To retrieve the type name without leading "?" the new
> ReflectionNamedType::getName() method can be used.

https://github.com/php/php-src/blob/9650b8e2412a919df8c5c626a1def937bc6d77d5/UPGRADING#L127-L129

Comment (iltar) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: In case of the `toString`, shouldn't the ? be stripped?

Comment (teohhanhui) [inline src/Symfony/Component/PropertyInfo/Extractor/ReflectionExtractor.php]: If there is a `getName` method, then we won't be using `__toString`, and if we are using `__toString` that means it's PHP <= 7.0 so there is no `?` prepended anyway. Stripping of characters is only a temporary solution anyway, as the `__toString` is likely to keep changing, judging from the mailing list discussion.

**Comment by dunglas** [comment]
Fabbot error is a false positive (it doesn't support PHP 7.1 yet).

👍 but I propose to wait and see what happens to this feature before the release of 7.1 to merge the PR.

**Comment by teohhanhui** [comment]
If we want people to test Symfony with PHP 7.1, we should provide the compatibility. It's already in RC now, so things shouldn't change as much.

**Comment by nicolas-grekas** [comment]
@teohhanhui see #19817 and related failure

**Comment by fabpot** [comment]
Thank you @teohhanhui.
